text/0103-introduce-core-index-commitment.md

```diff
@@ -0,0 +1,252 @@
+- the UMP queue contains a core index commitment and it matches the one in the descriptor.
+
+
+### UMP transport
+
+[CandidateCommitments](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/primitives/src/v7/mod.rs#L652) remains unchanged as we will store scale encoded `UMPSignal` messages directly in the parachain UMP queue by outputing them in the [upward_messages](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/primitives/src/v7/mod.rs#L682). 
+
+The UMP queue layout is changed to allow the relay chain to receive both the XCM messages and `UMPSignal` messages. An empty message (empty `Vec<u8>`) is used to mark the end XCM messages and the start of `UMPSignal` messages.
+
+This way of representing the new messages has been chosen over introducing an enum wrapper to minimize breaking changes of XCM message decoding in tools like Subscan for example. 
+
+Example: 
+```rust
+[ XCM message1, XCM message2, ..., EMPTY message, UMPSignal::CoreSelector ]
+```
+
+#### `UMPSignal` messages
+
+```rust
+/// An `u8` wrap around sequence number. Typically this would be the least significant byte of the parachain block number.
+pub struct CoreSelector(pub u8);
+
+/// An offset in the relay chain claim queue.
+pub struct ClaimQueueOffset(pub u8);
+
+/// Default claim queue offset
+pub const DEFAULT_CLAIM_QUEUE_OFFSET: ClaimQueueOffset = ClaimQueueOffset(1);
+
+pub enum UMPSignal {
+	/// A message sent by a parachain to select the core the candidate is commited to.
+	/// Relay chain validators, in particular backers, use the `CoreSelector` and `ClaimQueueOffset`
+	/// to compute the index of the core the candidate has commited to.
+	///
+	SelectCore(CoreSelector, ClaimQueueOffset),
+}
+```
+
+As we dont want to have a claim queue snapshot in the parachain runtime, we need to set `ClaimQueueOffset` 
+statically to some sane value. Parachains should prefer to have a static value that makes sense for their usecase which can be changed by governance at some future point. Changing the value dynamically can be a friction point. It will work out fine to decrease the value to build more into the present. But if the value is increased to build more into the future, a relay chain block will be skipped.
+
+Considering `para_assigned_cores` is a sorted vec of core indices assigned to a parachain at the
+specified claim queue offset, validators will determine the committed core index like this:
+
+```rust
+let assigned_core_index = core_selector % para_assigned_cores.len();
+let committed_core_index = para_assigned_cores[assigned_core_index];
+```
+
+
+### Polkadot Primitive changes
+
+#### New [CandidateDescriptor](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/primitives/src/v7/mod.rs#L482)
+
+- reclaim 32 bytes from `collator: CollatorId` and 64 bytes from `signature: CollatorSignature` and rename to `reserved1`  and `reserved2` fields.
+- take 1 bytes from `reserved1` for a new `version: u8` field.
```

Comment:
If the reserved fields are not zeroed, then we consider that the descriptor is v1 and expect valid signature. Nodes should check the signature in that case.

Reply:
It could, the `version` field is future stuff. We can only do this once we were able to fully deprecate/remove support for the legacy version having signatures and and ids. For launch of this we have to compare the entire signature field with `0`.